crank render - add support for observed connectionDetails

Avarei opened the pull request:
<!--
Thank you for helping to improve Crossplane! Please read the contribution docs
(linked below) if this is your first Crossplane pull request.
-->

### Description of your changes
Adds support for observed resources to use connectonDetails.

This allows adding kubernetes Secrets to observed resources in the `crossplane render` command. when another observed Resource targets the secret by having a spec.writeConnectionSecretToRef with the name and namespace of the kubernetes secret the secret data is set as the connectionDetails.

e.g.:
observed.yaml:
```yaml
---
apiVersion: test/v1
kind: xResource
metadata:
  annotations:
    crossplane.io/composition-resource-name: my-resource
spec:
  writeConnectionSecretToRef:
    name: foo
    namespace: bar
---
apiVersion: v1
kind: Secret
metadata:
  name: foo
  namespace: bar
data:
  foo: Zm9v
```
during the function run, the observed resource `my-resource` has the connectionDetails set based on the data of the secret.
<!--
Briefly describe what this pull request does, and how it is covered by tests.
Be proactive - direct your reviewers' attention to anything that needs special
consideration.

We love pull requests that fix an open issue. If yours does, use the below line
to indicate which issue it fixes, for example "Fixes #500".
-->


I have: <!--You MUST either [x] check or [ ] ~strike through~ every item.-->

- [x] Read and followed Crossplane's [contribution process].
- [x] Run `earthly +reviewable` to ensure this PR is ready for review.
- [x] Added or updated unit tests.
- [ ] Added or updated e2e tests.
- [ ] Linked a PR or a [docs tracking issue] to [document this change].
- [ ] Added `backport release-x.y` labels to auto-backport this PR.

Need help with this checklist? See the [cheat sheet].

[contribution process]: https://github.com/crossplane/crossplane/tree/main/contributing
[docs tracking issue]: https://github.com/crossplane/docs/issues/new
[document this change]: https://docs.crossplane.io/contribute/contribute
[cheat sheet]: https://github.com/crossplane/crossplane/tree/main/contributing#checklist-cheat-sheet

Avarei commented on cmd/crank/render/render.go:
```diff
@@ -74,9 +77,6 @@ type Inputs struct {
 	ExtraResources      []unstructured.Unstructured
 	Context             map[string][]byte
-
-	// TODO(negz): Allow supplying observed XR and composed resource connection
-	// details. Maybe as Secrets? What if secret stores are in use?
```
This solution does not add support for secret stores.

negz reviewed (commented):
Nice, thanks for tackling this!

Avarei commented on cmd/crank/render/render.go:
```diff
@@ -187,18 +187,52 @@ func Render(ctx context.Context, log logging.Logger, in Inputs) (Outputs, error)
+		var connectionDetails managed.ConnectionDetails
+		if connectionSecretRef := or.GetWriteConnectionSecretToReference(); connectionSecretRef != nil {
+			connectionDetails = observedConnectionDetails[*connectionSecretRef]
+		}
```
I could but I think it would be a little awkward since i cannot assign directly to a [field of a map](https://pkg.go.dev/golang.org/x/tools/internal/typesinternal#UnaddressableFieldAssign).

I would have to do something like this:
```go
if ref := or.GetWriteConnectionSecretToReference(); ref != nil {
	crs := observed[composite.ResourceName(name)]
	crs.ConnectionDetails = observedConnectionDetails[*ref]
	observed[composite.ResourceName(name)] = crs
}
```

but i tried to improve the spacing and variable names to make it easier to read.

jbw976 commented:
/fresh thanks for this contribution @Avarei! I think it's still relevant and hopefully we can finish it off in the v1.20 milestone 🙇

jbw976 commented on cmd/crank/render/render_test.go:
```diff
@@ -739,6 +739,207 @@ func TestRender(t *testing.T) {
+			},
+			want: want{
+				out: Outputs{
+					// TODO: expect connectionDetails in composite when support is added
```
@Avarei do you want to go ahead and tackle this part too? if i'm reading https://github.com/crossplane/crossplane/issues/4808 correctly, then it would be fully completed with what you've already done in this PR plus if you were to add connection details in the output. What do you think?

Avarei replied:
Here you go :)
I've implemented it in https://github.com/crossplane/crossplane/pull/5985/commits/c87edf5aba954ddaeed904fb2303302947378234.

Avarei replied:
I'm not sure how to add it as an e2e test, but I hope this helps: https://gist.github.com/Avarei/19bae623e8fe5290991698d4fb3cacd4

yohanb commented:
bump!  I've needed this feature and tried @avarei's fork and it works 👍
What's missing to get this merged?  @avarei if you want add me to your fork, I can help.

avarei commented:
Sorry for my late response.

@phisco I've reverted the change which added the XR support that @jbw976 requested

@negz I've rebased the branch

@yohanb I've invited you to the fork in case more requests come up

jbw976 commented:
@avarei, I was coming back to this PR now with v2.1 approaching, but immediately thought of all the confusion right now around XR connection details no longer being supported:

* https://github.com/crossplane/docs/issues/1001 

Can you explain the functionality and experience of this PR, e.g. by updating the description, in the context of removed functionality in v2? Does this PR still handle some supported scenarios? Is it intended to work with legacy XRs? something else?

Any extra context will be helpful here as so much has changed with connection details in v2 and it's not really well documented and understood by folks yet. Thanks for helping out to lessen the confusion here 🙇‍♂️
